### ISSUE TITLE
Proposal: Zero allocation connectionless sockets

### ISSUE THREAD

**Issue by scalablecory** [dotnet/runtime#30797]
This proposal eliminates allocations for connectionless use of `Socket`. It augments the `SocketAddress` class to allow reuse across operations, becoming a high-perf alternative to `EndPoint`.

# Rationale and Usage

APIs which need to translate between `IPEndPoint` and native `sockaddr` structures are performing a large amount of defensive copying and layering workarounds.

This affects UDP performance and contributes to excessive GC. A simple example is:

```c#
Socket socket = ...;
byte[] buffer = ...;
var remoteEndPoint = new IPEndPoint(IPAddress.Any, 0);

socket.ReceiveFrom(buffer, 0, buffer.Length, SocketFlags.None, ref remoteEndPoint);
socket.SendTo(buffer, 0, buffer.Length, SocketFlags.None, remoteEndPoint);
```

These two calls allocate 12 times:
* 3x `IPEndPoint`
* 3x `IPAddress`
* 3x `SocketAddress`
* 3x `byte[]` ...6x if IPv6

See also: dotnet/runtime#30196 

New usage has 0 allocations:

```c#
Socket socket = ...;
byte[] buffer = ...;
var remoteAddress = new SocketAddress(AddressFamily.InterNetwork);

socket.ReceiveFrom(buffer, 0, buffer.Length, SocketFlags.None, remoteAddress);
socket.SendTo(buffer, 0, buffer.Length, SocketFlags.None, remoteAddress);
```

# Proposed API

```c#
class Socket
{
	public int ReceiveFrom(Span<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress);
	public ValueTask<int> ReceiveFromAsync(Memory<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress, CancellationToken cancellationToken = default);

	public int SendTo(ReadOnlySpan<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress);
	public ValueTask<int> SendToAsync(ReadOnlyMemory<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress, CancellationToken cancellationToken = default);
}

class SocketAsyncEventArgs
{
	// Only one of RemoteEndPoint or RemoteAddress must be specified.
	public SocketAddress RemoteAddress { get; set; }
}

class SocketAddress
{
	// If we can merge System.Net.Primitives and System.Net.Sockets, these two methods are unnecessary. That would be ideal.
	public static void GetBuffer(SocketAddress address, out byte[] buffer);
	public static void SetSockaddrSize(SocketAddress address, int size);
}

class EndPoint
{
	public virtual void SerializeTo(SocketAddress socketAddress); // Already has "SocketAddress Serialize()"; default would be to call that and copy.
}
```

# Details

* It is intended that UDP servers will use `SocketAddress` as a dictionary key to lookup client state, to avoid first converting to `EndPoint`.
   * It is assumed that users will only rarely care to actually get the IP/port/etc. from the `SocketAddress`. This duty continues to be delegated to `EndPoint`.
   * If users ever want to deserialize a `SocketAddress` into an `EndPoint`, they can already use `EndPoint.Create`.
   * Need to ensure only the actual sockaddr structure is compared/hashed, not the entire byte buffer.
* `SocketAddress` is currently duplicated in System.Net.Primitives and System.Net.Sockets to avoid exposing its internal buffer. This change will allow avoiding duplication.
* This relies on users not using a `SocketAddress` until the I/O is finished.
   * This is a bit safer with `EndPoint` as we can take defensive copies before methods return.
* We can currently do some optimizations to avoid all allocations for `SendTo` and `SendToAsync` IPv4/IPv6 with some special casing, so this API would primarily be to optimize `ReceiveFrom` variants as well as (less important) allowing non-IPv4/IPv6 protocols to benefit. Still, if we were to add an API for `ReceiveFrom` we would probably want an API on `SendTo` for symmetry.

# Open Questions

* We've put some effort into not doing something like this before. It would be great to understand why. Currently:
	* We duplicate the `SocketAddress` class in multiple assemblies to avoid exposing its buffer, and have a step to marshal (byte-by-byte) between the two implementations.
	* Tons of APIs take `EndPoint`, it's a nice abstraction that we wanted here despite performance implications.
* It isn't immediately obvious from the API surface that `socketAddress` is written to by `ReceiveFrom`. Is there a better way we can indicate this?
* The two new methods on `SocketAddress` exist purely because System.Net.Sockets needs access to internals in System.Net.Primitives. Any thoughts on how to avoid exposing these "pubternal" bits?
   * One option is to merge System.Net.Primitives and System.Net.Sockets; I don't see harm in this but that is a much larger discussion :)
* If we merge the Primitives and Sockets assemblies, we can get rid of some of the allocations for `ReceiveFrom` without making any API changes. It's not a perfect solution but might be good enough.

# Related Issues

There are two additional issues to update our APIs with `ValueTask`/`Span`/`Memory` that this will need to be consistent with:
* Update `Socket` class dotnet/runtime#938 
* Update `UdpClient` class dotnet/runtime#864

**Comment by macaba** [comment]
@scalablecory What's required to progress this proposed API?

**Comment by scalablecory** [comment]
@macaba it will likely be required if we get a managed QUIC implementation. Otherwise, I'd love to see some good customer stories here.

**Comment by macaba** [comment]
I'm the architect of a virtual VHF radio simulation backend (for virtual air traffic control).

We typically get around 20k UDP packets per second (20-30Mbps).

There is a future feature upgrade coming in that will probably push this up to 50-70k UDP packets per second.

I'm looking to improve the efficiency as I want to keep the server cost increase to a minimum when packet counts go up.

**Comment by karaziox** [comment]
We have a service acting as a video stream redirector, receiving a stream from some network interface via some transport, and retransmitting it over another, possibly changing the transport.

Currently (on .Net Framework 4.8), we are fully limited by GC throughput at around 800 Mbps worth of stream redirection bandwidth. We are getting above the 50% time paused in GC, and we can't keep up with the incoming data, introducing packet losses. We have pooled pretty much everything under our control and 95%+ of the remaining allocations are caused by Socket. In our case we currently rely on the ReceiveFromAsync and BeginSend codepaths.

I understand some work in .Net Core was done that will certainly reduce those allocations, but the big per-receive allocations are still present: SocketAddress, IPEndPoint etc. Having a better way to do high throughput transmission on UDP sockets would help us a lot.

**Comment by ArnaudValensi** [comment]
Same thing here, we are developing a video game.

**Comment by rzikm** [comment]
I am implementing QUIC in C# and, truly most of the allocated memory occurs during `SendTo` and `ReceiveFrom` methods.

**Comment by mikebromwich** [comment]
Our use case - we have a managed implementation of SCTP - we've managed to reduce allocations down to the bare minimum, and the remaining allocations are almost all within Socket (which we're using in raw mode).

**Comment by nxrighthere** [comment]
If you want a temporary solution for UDP sockets, [here's](https://github.com/nxrighthere/NanoSockets) our abstractions over portable native API.

**Comment by asimonf** [comment]
I'm building a HID redirector for IOT purposes. My redirector is timesensitive, GC allocations are introducing more latency than necessary. I've already eliminated all allocations from other sources, so the remaining ones are strictly from the ReceiveFrom method. 

If modifying the API is too risky due to current users of the code, maybe integrating something different for zero allocations would be a good idea.

**Comment by antonfirsov** [comment]
Had a very brief look at the issue and the discussion, and didn't dig too deep into the code, so I may miss something, but I think for the synchronous `ReceiveFrom` defining an `EndPoint.SerializeTo(Span<byte>)` should be enough to avoid the `ReceiveFrom` allocations.

Currently, we are overriding `ref remoteEp` only if the addresses differ:
https://github.com/dotnet/runtime/blob/0f5b75344d5858d76da403735ee34c71d9d69d54/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L1774-L1779

Of course, this approach wouln't work with `async`. Instead of defining API-s which need special care (`ReceiveFrom` changing the `SocketAddress` content), maybe we should consider turning `EndPoint` to a "freezable" type so we can cache it?

Note that we already violated immutability rules in #39313 by no longer copying `_localEndPoint` in the `LocalEndPoint` getter, which we still haven't decided if it's a bug or not.

@geoffkizer do you think it would make sense to create an `EndPoint` freezability proposal for 7.0? I think it could be a low-effort, consistent way to avoid allocations here, with minimal behavioral breaking changes, which actually wouln't break any sane code.

**Comment by GSPP** [comment]
If `IPEndPoint` and `IPAddress` are already mutable, can we not pass in an existing instance?

I also wonder, why `IPAddress` stores IPv6 addresses as an `ushort[]`. Can we not store that as two `long` fields inline? The field is private to the class and I can't see how a public consumer could rely on it. For IPv4, this would increase the size of the class by 4 bytes (+16 for the address, -8 for the array and -4 for the IPv4 address field). For IPv6, it would reduce allocations by, I think, 24 bytes (no array allocation and object reference). The same remarks apply to the existing `SocketAddress` class.

**Comment by antonfirsov** [comment]
> For IPv4, this would increase the size of the class by 4 bytes (+16 for the address, -8 for the array and -4 for the IPv4 address field). For IPv6, it would reduce allocations by, I think, 24 bytes (no array allocation and object reference).

Since DualMode became more or less standard across our stack, client socket addresses are typically IPv6, so this sounds like a an overall win to me.

> If `IPEndPoint` and `IPAddress` are already mutable, can we not pass in an existing instance?

This is very similar @scalablecory's original proposal is. I'm skeptical whether such an approach meets our design/API guidelines.

**Comment by Turnerj** [comment]
Just to add another user/customer story to this, I'm working on an OSS low/no-allocation DNS client called [Dino DNS](https://github.com/TurnerSoftware/DinoDNS) and for a single query I'm seeing ~1700 bytes of allocations in my own benchmarks. I'm already pooling sockets and doing other tricks to minimise allocations and as far as I've been able to measure, the majority of those allocations left seem to be internal to both UDP (and TCP) socket reading/writing.

**Comment by vis2k** [comment]
Hey guys, great to see this issue being discussed. I'd like to give some feedback really quick.
For context, I am the developer of [Mirror](https://github.com/vis2k/Mirror), the largest open source game networking library for Unity.

For multiplayer games, we can not have runtime allocations.

- On clients, this causes noticeable GC freezes. This is especially troublesome in VR games where you want to avoid motion sickness :)
- For servers, we need to process thousands of UDP messages every update. Performance is already stretched to the very limit, we can not afford additional GC pauses.

For anyone using Unity, managed C# UDP sockets are essentially unusable.
- The most common solution is to ship a native C DLL for allocation free send/recv, which comes with a whole lot of headaches as you can imagine.
- To avoid having to build native DLLs on every platform, we managed to at least _reduce_ the allocations by using custom "Create(SocketAddress)", "Serialize", "GetHashCode" implementations. However, this only seems to work with Mono, which Unity is soon moving away from.

C# is an amazing(!) programming language.
Please do support allocation free UDP sockets, so that C# may become a viable option for multiplayer game netcode too.

This is extremely important for thousands of multiplayer game projects affecting millions of players :)

**Comment by c0nd3v** [comment]
@stephentoub We need you sir

**Comment by wfurt** [comment]
Is there some specific usage pattern @vis2k? For example I remember some parts where we try to capture the remote info but I'm not sure if that is really useful for games. We can probably also just go with the API as proposed. 

And I'm also wondering if Quic could be possible long term solution for you. While there is still space for performance and functional improvement I think is meant for cases like this. (just curious)

**Comment by FakeByte** [comment]
> Is there some specific usage pattern @vis2k? For example I remember some parts where we try to capture the remote info but I'm not sure if that is really useful for games. We can probably also just go with the API as proposed.
> 
> And I'm also wondering if Quic could be possible long term solution for you. While there is still space for performance and functional improvement I think is meant for cases like this. (just curious)

A lot of multiplayer games use unreliable send with UDP as it often makes no sense to receive old data. E.g. a players position does not need to be resend if a new position already arrived. Thats why I think Quic is not really an alternative to UDP for multiplayer games.

**Comment by vis2k** [comment]
> Is there some specific usage pattern @vis2k? For example I remember some parts where we try to capture the remote info but I'm not sure if that is really useful for games. We can probably also just go with the API as proposed.
> 
> And I'm also wondering if Quic could be possible long term solution for you. While there is still space for performance and functional improvement I think is meant for cases like this. (just curious)

Thanks for the quick reply.
I am no expert on .net internals, but if OP's proposal solves the SendTo / RecvFrom allocations then that's great :)

Again, thanks for considering this proposal!
This will be hugely valuable to the world of multiplayer games, it's difficult to imagine just how much of an impact this would have over the next decades.

**Comment by antonfirsov** [comment]
> I am no expert on .net internals, but if OP's proposal solves the SendTo / RecvFrom allocations then that's great :)

My conclusion from the looong discussion is that the OP proposal was controversial because it doesn't fit existing API patterns. There is a lot we can optimize without changing the API (see https://github.com/dotnet/runtime/issues/30797#issuecomment-530556553), we should start there.

**Comment by wfurt** [comment]
ok, that make sense @FakeByte. I'm trying to understand better the use case (again). For example you can use `Send` without even fiddling with addresses. What makes `SendTo` and `ReceiveFrom` attractive? Management of single socket? 

And how people would use the `SocketAddress`? Getting some specific use pattern and looking at general use case would be nice e.g. providing examples before jumping to conclusion.  

I think this lost traction after retreat of the original champions. But it is by far the most upvoted issue for `Sockets` and we may have capacity to address it in 8.0. The timing is also right as the planning just started and we should have enough time to workout the details.

**Comment by FakeByte** [comment]
> ok, that make sense @FakeByte. I'm trying to understand better the use case (again). For example you can use `Send` without even fiddling with addresses. What makes `SendTo` and `ReceiveFrom` attractive? Management of single socket?

Send would require a connected socket, which is not possible with udp. Having multiple sockets would have also no benefits with udp since sockets are not connected. Thats why SendTo and ReceiveFrom are used. Normal Receive works with udp, but that would mean you dont know who sent a message, including that info into the payload is not an alternitive to prevent easy tampering from the other end.

**Comment by Turnerj** [comment]
> Send would require a connected socket, which is not possible with udp

You actually can use `SendAsync` for UDP because you can call connect on a socket, even if it is counter intuitive to do so.

```csharp
var socket = new Socket(endPoint.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
socket.Connect(endPoint);
await socket.SendAsync(requestBuffer, SocketFlags.None);
```

I do this exact thing in a project of mine: https://github.com/TurnerSoftware/DinoDNS/blob/289a4e77384327defcdedbb94d2c94f48d8bb9d0/src/TurnerSoftware.DinoDNS/Connection/Resolvers/UdpResolver.cs#L7-L58

**Comment by c0nd3v** [comment]
> Send would require a connected socket, which is not possible with udp.

It’s definitely possible. We can use SocketOptionName.ReuseAddress to create a connected socket from an unconnected socket, which we can then use Send/Receive with. Unfortunately on Windows we don’t have the luxury luxury to use Receive in this manner, so we must use ReceiveFrom to identify the sender.

**Comment by c0nd3v** [comment]
@wfurt ReceiveFrom is primarily used to uniquely identify where the data is coming from.

**Comment by FakeByte** [comment]
> > Send would require a connected socket, which is not possible with udp.
> 
> It’s definitely possible. We can use SocketOptionName.ReuseAddress to create a connected socket from an unconnected socket, which we can then use Send/Receive with. Unfortunately on Windows we don’t have the luxury to use Receive in this manner, so we must use ReceiveFrom to identify the sender.

This is correct, windows will not hand the udp packet back to the connected socket, but instead it will give the packet to any socket that is listening to a specific port.

**Comment by sgf** [comment]
Network programming is an important cornerstone of a language. Now that people have discovered and proposed improvements, can the official solve these problems as soon as possible. Or if the official has a better idea, please provide it as soon as possible.
If such things are ignored for several years, it will be difficult for developers to make substantial progress in their projects.
I have been engaged in C# development for many years, and I have always known that C#'s network programming performance itself is relatively backward, and even if SocketEventArgs is used, the difficulty of writing is relatively high.

**Comment by davidfowl** [comment]
> I have been engaged in C# development for many years, and I have always known that C#'s network programming performance itself is relatively backward, and even if SocketEventArgs is used, the difficulty of writing is relatively high.

While this is true, the core streaming socket APIs have been allocation free (amortized) since .NET Core 3.1 https://github.com/davidfowl/DotNetCodingPatterns/blob/main/2.md.

TCP has gotten the Lionshare of the performance optimizations for obvious reasons 😄 .

**Comment by sgf** [comment]
> > I have been engaged in C# development for many years, and I have always known that C#'s network programming performance itself is relatively backward, and even if SocketEventArgs is used, the difficulty of writing is relatively high.
> 
> While this is true, the core streaming socket APIs have been allocation free (amortized) since .NET Core 3.1 https://github.com/davidfowl/DotNetCodingPatterns/blob/main/2.md.
> 
> TCP has gotten the Lionshare of the performance optimizations for obvious reasons 😄 .



In fact, no matter whether it is .net or windows, even the Tcp protocol optimized by the .net team is still not good enough. Because in fact, Google invested more in the TCP protocol, although I heard that BBR may not be directly ordered by Google, but It is a private project.
But BBR does solve very large problems to a large extent.
Of course, TCP is a very old protocol after all, no matter how optimized it is, it will not help, so people are launching many protocols now, and most of these new protocols are built on top of the UDP protocol. For example, quic and so on.

I am sure that the UDP protocol has not received real attention for a long time, whether it is in the Windows system or the .net class library. Its performance on Windows is only a small part of the Linux system. It was not until recently that Windows Server 2022/Win11 has little improved, However, this means that everything is left behind.

I am very happy that it seems that the current issue has been taken seriously, but I am skeptical whether it will be able to get real improvements in the end. I have been in the github community for many years, and i know ever there are many features that even if they get the champion, even if the prototype is implemented, they are not actually implemented. Not merged into the trunk, not released with the standard library, and .net team ignored them.

Good luck with this feature (improvement).

**Comment by wfurt** [comment]
plan was discussed for 8.0. It is unclear if we have enough resources to get it done. 
Obviously we cannot do better than OS but it seems like most of the issues comes from .NET layer. 
It would be great if we can line up some users for alfa/beta testing.

**Comment by friflo** [comment]
@sgf 

> ... most of these new protocols are built on UDP protocol. For example, quic and so on.

Absolutely correct observation.

**Comment by davidfowl** [comment]
Unlike TCP we don't have any protocol implementations in .NET that can be used to drive the performance work. A managed QUIC implementation would do, or some other protocol, like WebRTC would be good.

> It would be great if we can line up some users for alfa/beta testing.

As @wfurt said, it would be great to drive this work with something other than just raw udp socket benchmarks....

**Comment by Turnerj** [comment]
I'm happy to alpha/beta test solutions for my DNS library (https://github.com/TurnerSoftware/DinoDNS) as I've been trying to have extremely low/no allocations throughout it.

**Comment by vis2k** [comment]
Hello friends, checking in a few months later - great to see some progress here.
Sharing a few more experiences with high throughput C# UDP, for those who care.

- Working on a project which requires multiple non-blocking UDP sockets in the same thread. Epoll/kqueue is needed to quickly check which of the sockets are readable/writable. Picked C# as first choice as usual. 
  - C#'s Select() method supposedly uses epoll/kqueue, even though the documentation doesn't make it obvious. Works as expected, everything was great. Did a Rust implementation for comparison, and ended up with 5x the throughput. 80 MB/s for C# SendTo/ReceiveFrom/Select, and 400 MB/s with Rust. Expected maybe 30-50% worse performance, that would've been just fine considering how C# is overall way more pleasant. However, a factor of 5x is just way too much, so I had to choose Rust over C#
- Recently switched from blocking to non-blocking UDP sockets for Mirror (Unity networking). Socket.SendTo/ReceiveFrom throws WouldBlock if it would block, which is to be expected. Unfortunately a user reported that C# allocates new memory every time it throws the WouldBlock exception (https://github.com/MirrorNetworking/Mirror/issues/3413). Which means we need to use Poll() again, even though the socket is already non-blocking. This would be unnecessary in most other languages.

C# UDP is stable and has all the features, which is great!
Unfortunately everything seems to allocate, and/or is super slow.

Just to be clear, we love C# and want to use it for everything ❤️.
The disappointment here comes from having to switch to Rust/Go for UDP, when we really want C# :)

**Comment by antonfirsov** [comment]
Honestly, with the current team load I believe it's impossible to get this done in a "zero allocation" manner for .NET 8.0, since that would require us to either push through unconventional API ideas (like the one proposed by the OP), or to implement a complicated caching mechanism for `IPEndPoint` instances in `ReceiveFrom`.

Instead of chasing an overly ambitious goal, and keep postponing a huge one-time investment from release to release, I wonder how valuable would it be for the community to make this at least *somewhat* better first, ie. gradually reducing but not fully eliminating allocations with improvements proposed in https://github.com/dotnet/runtime/issues/30797#issuecomment-530556553. Let's assume we manage to reduce allocations by an amount of 2x, and improve our throughput as a consequence, but there would be still gaps compared to other stacks. Would be .NET still no-go for any serious UDP application? Thoughts?

**Comment by sgf** [comment]
To be honest, I suggest to promote it and implement it as soon as possible, because according to my understanding, once it is shelved, it will be unknown when it will be restarted. And from the perspective of the problem, this is obviously a relatively simple problem, isn't it.

**Comment by sgf** [comment]
Due to the simplicity of the udp protocol itself, its api itself will not change significantly in terms of external calls. If there are other internal performance problems, it can be further optimized, but the current problem seems to be obvious. So make sure that the external api remains , to solve the current allocation problem caused by the ip address. Subsequent optimization issues in other aspects are not within the scope of the current issue. After all, the current issue only discusses allocation.

**Comment by FakeByte** [comment]
If I need UDP sockets in c# I always use a wrapper around native sockets, for me a "quick win" solution would not be enough to convince me to switch to using the built in socket api, I would assume for most high throughput applications this would be the same.

**Comment by wfurt** [comment]
It would be great IMHO to write benchmarks for scenarios people care about https://github.com/dotnet/performance
And yes, throwing is expensive. If there is workflow with it that is interesting there should be separate benchmark for it.
This is one area where the community can easily help us.

**Comment by sgf** [comment]
> If I need UDP sockets in c# I always use a wrapper around native sockets, for me a "quick win" solution would not be enough to convince me to switch to using the built in socket api, I would assume for most high throughput applications this would be the same.


I have nothing against you wrapping the platform api yourself. 
  But everyone encapsulates the platform API, which violates the DRY principle.that waste resource of the worldThis is not environmentally friendly. Increased carbon emissions.
 The significance of .net cross-platform is to smooth out platform differences. Our requirements for the .net platform are not high. There are many network protocols, but there are only two basic protocols, one is TCP and the other is UDP. The UDP protocol is even as simple as send and receive, not listen and connect.

**Comment by c0nd3v** [comment]
@sgf What the hell are you rambling about? Are you seriously trying to say that writing a wrapper around native sockets causes pollution?

Please stop spreading misinformation there is people here who are trying to get work done and need to write high-throughput UDP applications in C#

**Comment by mikebromwich** [comment]
I am a huge fan of C#/.NET - it is definitely our go-to language for the majority of the work we do. The performance achievable is really impressive for such a rich and highly abstracted language, and particularly for a garbage-collected language.

It is however, just that - a garbage-collected language - and so avoiding allocations/garbage if never going to be easy. For some of the use cases mentioned, if consistent, very high performance and tail latencies are critical, I wonder whether C# is the best choice? In our telecoms and networking work, we use C# for our control plane - great to tackle complexity, and Rust for our data/user plane - great for performance (closer to the wire with no GC). There are many IPC or RPC mechanisms that can be used to bridge between the two - and so harness the benefits of both.

**Comment by FakeByte** [comment]
> Of course, it is undeniable that his intention may just be to show that the built-in API is really not good enough. So much so that he had to choose to wrap the Api himself. If I misunderstood what he meant, then I apologize for that.

Thats what I meant, using a native wrapper is of course not an ideal solution, needing to compile the native dll for each platform separately is very annoying. But for the performance it is worth it at the moment.

**Comment by Turnerj** [comment]
> for me a "quick win" solution would not be enough to convince me to switch to using the built in socket api

Just to be clear with my "quick win" comment, I'm referring to @antonfirsov's comment of work load available for changes in .NET 8. I'd rather even some improvements that we can get in this coming release than just waiting till some release in the future where all the changes required can be tackled at once.

Whatever changes that can be made to .NET 8 given the workload might not be enough to get you to switch to using the managed implementation but it will benefit everybody already using it. When future releases contain more improvements, it may still get to a point where you actually can with minimal or no performance degradation.

Really, all I'm talking about is time to get some changes in now and some more later rather than getting them all at once.

> But for the performance it is worth it at the moment.

I am kinda curious how much better throughput/lower latency you're actually finding the native sockets over the .NET implementation. Is it like 2x faster or 10x faster? More? This is relating back to @antonfirsov's comment, if the managed implementation was 2x faster/half the allocations than it is currently, does that even remotely close the gap?

**Comment by azureskydiver** [comment]
As much as I also agree that getting any quick wins as soon as possible is a great idea, I think some people are discounting the time and effort it will take for some people to switch from their current set of wrappers back to the .NET APIs. If the custom wrapper was written with a completely different API style/approach than the .NET version, obviously it is not a drop in replacement, nor will it be a "simple refactor" . 
And after all the effort to switch back to the .NET version, you'll need to do a full set of benchmarking and profiling again to prove to management that it was worth that time and and effort to do the switch back, and to throw away the previous custom work. Furthermoree, you can't just tell your test team/organization that "it's okay, you don't need to do a full regression test".

**Comment by sgf** [comment]
> Thats what I meant, using a native wrapper is of course not an ideal solution, needing to compile the native dll for each platform separately is very annoying. But for the performance it is worth it at the moment.

Well, I misunderstood you, and I apologize for that.

**Comment by sgf** [comment]
https://github.com/enclave-networks/Enclave.FastPacket
https://github.com/JohannesDeml/NetworkBenchmarkDotNet
https://github.com/enclave-networks/research.udp-perf

here is some repositories maybe little help

**Comment by vis2k** [comment]
Hello friends!
Another day in the UDP trenches.
Inspecting IPAddress.GetHashCode for allocations:
```cs

        public override int GetHashCode() {
            //
            // For IPv6 addresses, we cannot simply return the integer
            // representation as the hashcode. Instead, we calculate
            // the hashcode from the string representation of the address.
            //
            if ( m_Family == AddressFamily.InterNetworkV6 ) {
                if ( m_HashCode == 0 )
                    m_HashCode = StringComparer.InvariantCultureIgnoreCase.GetHashCode(ToString());

                return m_HashCode;
            }
            else {
                //
                // For IPv4 addresses, we can simply use the integer
                // representation.
                //
                return unchecked((int)m_Address);
            }
        }
```

This function allocates because address byte[] is converted to a string(!), then hashed.
It's completely unnecessary to allocate a StringBuilder, copy the byte[] into the string, then hash the string.
Why not just hash the byte[]. It's faster, it doesn't allocate, and it requires no extra features other than someone to care.

C# is one of the most popular programming languages in the world.
Netcode is hard, it's totally understandable how someone would take the easy shortcut with ToString here.

But this is not what you would expect from Microsoft :(

**Comment by wfurt** [comment]
I have PoC for the 
```c#
class Socket
{
	public int ReceiveFrom(Span<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress);
	public ValueTask<int> ReceiveFromAsync(Memory<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress, CancellationToken cancellationToken = default);

	public int SendTo(ReadOnlySpan<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress);
	public ValueTask<int> SendToAsync(ReadOnlyMemory<byte> buffer, SocketFlags socketFlags, SocketAddress socketAddress, CancellationToken cancellationToken = default);
}
```

The `socketAddress` is updated in place - assuming it was created for correct address family and has sufficient capacity. If not, the call would fail e.g. there is no attempt to resize the underlying buffer. And of course you cannot pass same instance to multiple calls.  

From the conversation above: for the DNS server:
https://github.com/TurnerSoftware/DinoDNS/blob/main/src/TurnerSoftware.DinoDNS/Connection/Listeners/UdpQueryListener.cs
One can reduce allocations significantly by using the new API directly and allocating `SocketAddress` inside of the receive loop. It cannot be above the loop as the code does not `await` `HandleRequestAsync`.  Probably best option would be to make it part of the `transitData`.  Aside from that the `SocketAddress` is used as opaque object so it should just work. 

The Mirroring is somewhat more complicated: https://github.com/MirrorNetworking/Mirror/blob/d226d577c21e557ae09086e0ddc1da63d2704e38/Assets/Mirror/Transports/KCP/kcp2k/highlevel/KcpServer.cs#L149

If the `RawReceiveFrom` is not called concurrently, one can simply create global `SocketAddress` for  IPv6 - it is ok to pass bigger buffer - my assumption is that the receive call would set `SocketAddress.Size` to reflect the actually received amount e.g. v4 vs v6 or file length for UDS. The connectionID seems to be simple `GetHashCode`. That should work on `SocketAddress` as well - with the caveat that of course different endpoint can possibly produce same hash code. (but that problem already exists)  #86872 would make it easier to do extra validation but it feels like it is not necessary strictly speaking. 

At the moment I don't have the originally proposed change for `SocketAsyncEventArgs`. It is not too difficult but for now I did not do it to keep the scope as small as possible. The work is mostly around validation as the field is already used internally and we would need to define and verify behavior for cases when  somebody would also specify `RemoteEndPoint `. 

At least for projects discussed above this should be sufficient. If this is good enough for majority here I can drive it for API review and push it to 8.0 before the gates close. 

I opened  separate #86872 to provide some more convenient methods around `SocketAddress`  but the goal goes beyond UDP. @antonfirsov opened https://github.com/dotnet/runtime/issues/78993 and that should make it easier for anybody who want's to interact with (IP)EndPoint as some point. While both are somewhat related, we may or may not get them into 8.0. As the work is not excessive I'd be happy to tentatively push both to 8.0 and ditch them if we run out of time. 

Lastly #86513 talks about cost of pinning. We had case when the `GCHandles` fragmented memory and caused OOM. Hopefully that should not be problem for the synchronous version but we talk with @stephentoub about possibility to use `NativeMemory` in some cases to remove  work from GC.

**Comment by FakeByte** [comment]
@wfurt 
The API you proposed looks really nice, changing socket address in place is a simple solution and solves the allocation problems.

**Comment by vis2k** [comment]
@wfurt That's amazing. This will elevate C# to a whole new class of use cases.
Super excited about this, and it's great to see that C# devs care so much.

**Comment by 32haojiufangjia** [comment]
什么时候可以完成这项0GC的工作啊？？？期待

**Comment by sgf** [comment]
it seems blocked

https://github.com/dotnet/runtime/issues/87397
https://github.com/dotnet/runtime/issues/86872

i think thats won't be release with .net 8.0

**Comment by antonfirsov** [comment]
Note that the label is `blocking`, not "blocked". This means that the issue is high-priority (blocking the release) which pushes it to the top of the API review backlog.

**Comment by sgf** [comment]
> Note that the label is , not "blocked". This means that the issue is high-priority (blocking the release) which pushes it to the top of the API review backlog.`blocking`

👍  great.

**Comment by wfurt** [comment]
This was long journey but the end is near. #87397 added sync & async overloads with `SocketAddress`. It is possible now to pass it to SendTo & ReceiveFrom in and avoid allocations. Same object will be updated in place e.g. it cannot be shared by multiple threads. On receive, the `Size` is updated to reflect valid bytes e.g. one will need to slice the `Buffer` to get valid address. #89808 should help on Windows with avoiding GCHandle & pinning. https://github.com/dotnet/runtime/pull/90086 also reduce allocation for legacy SendTo & Receive but there is still room for improvements in 9.0. 

What missing is direct support for `SocketAsyncEventArgs`. There did not seems to be any compelling reasons besides completeness. And #86872 did not get the helper functions approved. If anybody needs API to extract useful data from the `SocketAddress` please comment here or on  #86872. 

All this should be available ion RC1 build -> just barely made it in for 8.0. 
Since changes above should fix most use cases I'm going to close this issue. It would be great if people can pick up RC1 and experiment as everything got in pretty late.

**Comment by macaba** [comment]
Thank you, it's appreciated.

Should I use `new SocketAddress(AddressFamily.InterNetworkV6);` to ensure the internal buffer is large enough for a ReceiveFrom that could receive both V4 and V6?

**Comment by wfurt** [comment]
yes, it should have at least what the given socket can receive. Or more generically 
```c# 
Socket socket = new Socket(SocketType.Dgram, ProtocolType.Udp)
SocketAddress socketAddress = new SocketAddress(socket.Address.Family);
```

it will default to `DualMode` when it can and fall-back to IPv4 if needed. 
One more note that the the old `ReceiveFrom` would do work to map IPv4 to IPv6V4Mapped. 
The new API leaves the buffer exactly as given by OS. e.g. if you do `EndPoint.Create(socketAddress)` you may end up with different address and if you care you need to do the mapping yourself.
